File store optimizations

jonas opened the pull request:
## What changes are proposed in this pull request?

Optimize `mlflow/store/tracking/file_store.py` to do less I/O per API call.
For example `log_metric` has been changed to no longer load all metrics, params, and tags before writing new metric values.

Flamegraphs (generated with `py-spy`) for each commit in this PR can be found here: https://github.com/ElementAI/mlflow/tree/file-store-optimizations-svgs/bench-env/svgs

## How is this patch tested?

```
./lint.sh
./travis/run-small-python-tests.sh
```

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Optimize the file store to load less data from disk for each operation.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [x] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

---

jonas committed:
FileStore: optimize log_* and set_tag functions

Reduce I/O by only loading the RunInfo when possible.

jonas committed:
FileStore: optimize search_runs to operate on paths

Use the resolved run directory paths to reduce the amount of I/O.

jonas committed:
FileStore: load metrics, params and tags from run info

Reduce repeated calls to _get_run_info by introducing new internal
methods that operate on RunInfo.

jonas commented:
Updated to fix linting errors and make the validation checks more consistent.

I've tested the patch in an export script and the speed-up is quite noticeable. Similarly, `mlflow server` loads data significantly faster, i.e. no gunicorn worker timeouts even for experiments with a lot of runs.

jonas commented:
The Python Spark tests failed. Let me know if I should trigger a rebuild:
https://travis-ci.org/mlflow/mlflow/jobs/640942675#L1900

tomasatdatabricks reviewed (commented):
Looks good! Thanks for the contribution @jonas 

Linter found 2 issues, lgtm after those are addressed.

jonas commented:
@tomasatdatabricks Thanks for the review

I addressed the issues found in https://travis-ci.org/mlflow/mlflow/jobs/640645112 this morning
The resulting change was this: https://github.com/mlflow/mlflow/compare/eb4271e5cbe1666b26d0dc8c065b4eadf7e7e025..11dcaf53f97bec634c883b445265129515b7c61c

I will go ahead and trigger another build.

jonas committed:
Trigger CI

tomasatdatabricks reviewed (approved):
lgtm! Thx @jonas
